dart/common/SharedLibrary.hpp

```diff
@@ -0,0 +1,114 @@
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
+ *   CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+ *   INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+ *   MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ *   DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
+ *   CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ *   SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ *   LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
+ *   USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ *   AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *   LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *   ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *   POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef DART_COMMON_SHAREDLIBRARY_HPP_
+#define DART_COMMON_SHAREDLIBRARY_HPP_
+
+#include <memory>
+#include <string>
+#include "dart/common/Platform.hpp"
+
+#if DART_OS_LINUX
+#define DYNLIB_HANDLE void*
+#elif DART_OS_MACOS
+#define DYNLIB_HANDLE void*
+#elif DART_OS_WINDOWS
+#define DYNLIB_HANDLE hInstance
+#endif
+
+namespace dart {
+namespace common {
+
+namespace detail {
+class SharedLibraryManager;
+} // namespace detail
+
+/// SharedLibrary is a RAII object wrapping a shared library.
+class SharedLibrary
+{
+protected:
+  enum ProtectedContructionTag
+  {
+    ProtectedConstruction
+  };
+
+public:
+  /// Creates a SharedLibrary.
+  ///
+  /// \param[in] The filename of the shared library. If the filename doesn't
+  /// include the extension, this function will use the best guess depending on
+  /// the OS (e.g., '.so' for Linux, '.dylib' for macOS, and '.dll' for
+  /// Windows).
+  /// \return Pointer to the created SharedLibrary upon success in loading.
+  /// Otherwise, returns nullptr.
+  static std::shared_ptr<SharedLibrary> create(const std::string& fileName);
+
+  /// Constructs from a filename.
+  ///
+  /// \note SharedLibrary must be constructed by create().
+  ///
+  /// \param[in] The filename of the shared library. If the filename doesn't
+  /// include the extension, this function will use the best guess depending on
+  /// the OS (e.g., '.so' for Linux, '.dylib' for macOS, and '.dll' for
+  /// Windows).
+  /// \return Pointer to the created SharedLibrary upon success in loading.
+  /// Otherwise, returns nullptr.
+  explicit SharedLibrary(ProtectedContructionTag, const std::string& fileName);
+
+  /// Destructor
+  virtual ~SharedLibrary();
+
+  /// Returns the file name of the shared library
+  const std::string& getFileName() const;
+
+  /// Returns true if the shared library loading was successful.
+  bool isValid() const;
+
+  /// Returns a symbol from the shared library
+  void* getSymbol(const std::string& symbolName) const;
```

Comment:
What does it return if the symbol does not exist? `nullptr`?

Also, it might be worth mentioning that you have to `reinterpret_cast` the return value to the appropriate type to make use of the `void*` returned by this function.